### ISSUE TITLE
otp 24.0.6 process_info messages segfault 

### ISSUE THREAD

**Issue by hizel** [erlang/otp#5221]
**Describe the bug**
In last OTP 24.0.6 `process_info(Pid, messages).` segfault when in messages queue signal from monitor with custom tag

**To Reproduce**
tst2.erl
```erlang
-module(tst2).
-mode(compile).
-compile(export_all).

%tst2:main("hi1", "hi2","hi3","hi4","hi5","hi6").
main(Names) ->
  F = fun() ->
    Refs = [begin 
      Pid = spawn(?MODULE, worker, [list_to_binary(Name), 0]),
      Ref = erlang:monitor(process, Pid, [{tag, {worker_down, list_to_binary(Name), restart}}]),
      Ref
      end|| Name <- Names],
  
    loop(Refs, 0)
  end,
  spawn(F), 
  spawn(F),
  spawn(F),
  spawn(F),
  spawn(F),
  spawn(F),  
  ok.



worker(Name, C) ->
  %io:format("work ~s ~p~n", [Name, C]),
  if
    C rem 10 =:= 1 -> exit(olo);
    true -> ok
  end,
  %timer:sleep(10),
  ok.


loop(Refs, C) ->
  receive
    {{worker_down, Name, restart}, Ref, process, _, Why} ->
      Refs1 = lists:delete(Ref, Refs),
      case Why of
        olo -> io:format("~p~n", [{Name, C}]);
        _ -> ok
      end,
      Pid = spawn(?MODULE, worker, [Name, C]),
      Ref1 = erlang:monitor(process, Pid, [{tag, {worker_down, Name, restart}}]),
      loop([Ref1|Refs1], C+1)
  end.
```

```
erlc tst2.erl
erl +S 2:2 +A 1:1 -run tst2 main hi1 hi2 hi3 hi4 hi5 hi6 -sname tst2
```

separate console
```
erl -remsh tst2
```
find process
```
<0.85.0>              erlang:apply/2                         610 89296624    6
                      tst2:loop/2                              6              
<0.86.0>              erlang:apply/2                         610 12396621    6
                      tst2:loop/2                              6              
<0.87.0>              erlang:apply/2                        2586 11558621    6
                      tst2:loop/2                              6              
<0.88.0>              erlang:apply/2                         610   346942    6
                      tst2:loop/2                              6              
<0.89.0>              erlang:apply/2                        2586 11731506    6
                      tst2:loop/2                              6              
```

```
(tst2@core-hizel)2> process_info(<0.89.0>).
[{current_function,{tst2,loop,2}},
 {initial_call,{erlang,apply,2}},
 {status,waiting},
 {message_queue_len,6},
 {links,[]},
 {dictionary,[]},
 {trap_exit,false},
 {error_handler,error_handler},
 {priority,normal},
 {group_leader,<0.71.0>},
 {total_heap_size,3581},
 {heap_size,2586},
 {stack_size,6},
 {reductions,11731507},
 {garbage_collection,[{max_heap_size,#{error_logger => true,kill => true,size => 0}},
                      {min_bin_vheap_size,46422},
                      {min_heap_size,233},
                      {fullsweep_after,65535},
                      {minor_gcs,19714}]},
 {suspending,[]}]
(tst2@core-hizel)3> process_info(<0.89.0>, messages).
*** ERROR: Shell process terminated! (^G to start new job) ***
```

tst2  vm core dumps
```
size_object: bad tag for 0xc0
Aborted (core dumped)
```

```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:49
#1  0x00007ff446d8453a in __GI_abort () at abort.c:79
#2  0x0000559bab667cea in erts_exit_epilogue () at beam/erl_init.c:2534
#3  0x0000559bab738bfa in erts_exit_vv (n=-2, flush_async=flush_async@entry=0, fmt=fmt@entry=0x559bab9995df "size_object: bad tag for %#x\n", args1=args1@entry=0x7ff403db7478, args2=args2@entry=0x7ff403db7490) at beam/erl_init.c:2520
#4  0x0000559bab738d61 in erts_exit (n=n@entry=-2, fmt=fmt@entry=0x559bab9995df "size_object: bad tag for %#x\n") at beam/erl_init.c:2544
#5  0x0000559bab76efec in size_object_x (obj=<optimized out>, litopt=litopt@entry=0x0) at beam/copy.c:240
#6  0x0000559bab7b93aa in erts_proc_sig_prep_msgq_for_inspection (c_p=c_p@entry=0x559bac3353e8, rp=rp@entry=0x559bac3353e8, rp_locks=rp_locks@entry=1, info_on_self=info_on_self@entry=0, mip=mip@entry=0x7ff406cf0170) at beam/erl_proc_sig_queue.c:7032
#7  0x0000559bab74604b in process_info_aux (c_p=0x559bac3353e8, hfact=0x7ff403db79c0, rp=0x559bac3353e8, rp_locks=1, item_ix=4, flags=<optimized out>, reserve_sizep=0x7ff403db7978, reds=0x7ff403db79a8) at beam/erl_bif_info.c:1473
#8  0x0000559bab7465f3 in erts_process_info (c_p=c_p@entry=0x559bac3353e8, hfact=hfact@entry=0x7ff403db79c0, rp=rp@entry=0x559bac3353e8, rp_locks=rp_locks@entry=1, item_ix=item_ix@entry=0x559bac214244, item_ix_len=1, flags=61, reserve_size=<optimized out>, reds=0x7ff403db79a8) at beam/erl_bif_info.c:1022
#9  0x0000559bab7adb6c in handle_process_info (c_p=c_p@entry=0x559bac3353e8, tracing=tracing@entry=0x7ff403db7b20, sig=sig@entry=0x559bac2141b8, next_nm_sig=next_nm_sig@entry=0x559bac335550, is_alive=is_alive@entry=1) at beam/erl_proc_sig_queue.c:4193
#10 0x0000559bab7b514e in erts_proc_sig_handle_incoming (c_p=c_p@entry=0x559bac3353e8, statep=statep@entry=0x7ff403db7bf0, redsp=redsp@entry=0x7ff403db7bf4, max_reds=100, local_only=<optimized out>) at beam/erl_proc_sig_queue.c:5420
#11 0x0000559bab683e80 in erts_schedule (esdp=<optimized out>, p=0x559bac3353e8, calls=0) at beam/erl_process.c:9922
#12 0x00007ff4047e4238 in global::process_main ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
(gdb) frame 6
#6  0x0000559bab7b93aa in erts_proc_sig_prep_msgq_for_inspection (c_p=c_p@entry=0x559bac3353e8, rp=rp@entry=0x559bac3353e8, rp_locks=rp_locks@entry=1, info_on_self=info_on_self@entry=0, mip=mip@entry=0x7ff406cf0170) at beam/erl_proc_sig_queue.c:7032
7032                    Uint sz = size_object(msg);
(gdb) etp c_p
#Cp<0x559bac3353e8>.
(gdb) etp-process-info c_p
  Pid: <0.89.0>
  State: running-sys | active-sys | prq-prio-normal | usr-prio-normal | act-prio-normal

  Flags: 
  Current function: tst2:loop/2
  I: #Cp<tst2:loop/2+0x74>
  Heap size: 2586
  Old-heap size: 987
  Mbuf size: 8
  Msgq len: 6 (inner=6, outer=0)
  Msgq Flags: delayed-sigq-len handling-sig on-heap 
  Parent: <0.9.0>
  Pointer: (Process*)0x559bac3353e8
```

**Expected behavior**

**Affected versions**
OTP 24.0.6 jit x86_64
```
$ otp-OTP-24.0.6/bin/erl
Erlang/OTP 24 [erts-12.0.4] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [jit]
```

**Additional context**
Also tst3.erl with monitor without tag works well

```erlang
-module(tst3).
-mode(compile).
-compile(export_all).

%tst2:main("hi1", "hi2","hi3","hi4","hi5","hi6").
main(Names) ->
  F = fun() ->
    Refs = [begin 
      Pid = spawn(?MODULE, worker, [list_to_binary(Name), 0]),
      Ref = erlang:monitor(process, Pid),
      {Ref, list_to_binary(Name)}
      end|| Name <- Names],
  
    loop(Refs, 0)
  end,
  spawn(F), 
  spawn(F),
  spawn(F),
  spawn(F),
  spawn(F),
  spawn(F),  
  ok.



worker(Name, C) ->
  if
    C rem 10 =:= 1 -> exit(olo);
    true -> ok
  end,
  %timer:sleep(10),
  ok.


loop(Refs, C) ->
  receive
    {'DOWN', Ref, process, _, Why} ->
      {value, {Ref, Name}, Refs1} = lists:keytake(Ref, 1, Refs),
      case Why of
        olo -> io:format("~p~n",[{Name, C}]);
        _ -> ok
      end,
      Pid = spawn(?MODULE, worker, [Name, C]),
      Ref1 = erlang:monitor(process, Pid),
      loop([{Ref1, Name}|Refs1], C+1)
  end.
```

```
(tst3@core-hizel)3> process_info(<0.86.0>, messages).
{messages,[{'DOWN',#Ref<0.867995495.2987917313.110696>,
                   process,<0.7722.823>,normal},
           {'DOWN',#Ref<0.867995495.2987917313.110708>,process,
                   <0.7865.823>,normal},
           {'DOWN',#Ref<0.867995495.2987917313.110709>,process,
                   <0.7803.823>,normal},
           {'DOWN',#Ref<0.867995495.2987917313.110710>,process,
                   <0.7789.823>,normal},
           {'DOWN',#Ref<0.867995495.2987917313.110711>,process,
                   <0.7782.823>,normal}]}
```

**Comment by rickard-green** [comment]
Thanks for the detailed bug report! I've made a PR #5232 (based on OTP 24.0.6) which should fix this bug. It has only passed basic testing yet, though.

**Comment by hizel** [comment]
looks like works for me, thanks!

**Comment by rickard-green** [comment]
Now fixed in the just released [OTP 24.1.1](https://erlang.org/download/OTP-24.1.1.README) patch.
